python/src/com/jetbrains/python/console/PythonConsoleView.java

```diff
@@ -293,10 +300,78 @@ public void inputRequested() {
     splitWindow();
   }
 
+  protected final void doAddPromptToHistory(boolean isMainPrompt) {
+    flushDeferredText();
+    EditorEx viewer = getHistoryViewer();
+    DocumentEx document = viewer.getDocument();
+    RangeHighlighter highlighter = getHistoryViewer().getMarkupModel()
+      .addRangeHighlighter(document.getTextLength(), document.getTextLength(), 0, null, HighlighterTargetArea.EXACT_RANGE);
+    final String prompt;
+    if (isMainPrompt) {
+      prompt = myPromptView.getMainPrompt();
+      print(prompt + " ", myPromptView.getPromptAttributes());
+    }
+    else {
+      prompt = myPromptView.getIndentPrompt();
+      //todo should really be myPromptView.getPromptAttributes() output type
+      //but in that case flushing doesn't get handled correctly. Take a look at it later
+      print(prompt + " ", ConsoleViewContentType.USER_INPUT);
+    }
+
+    highlighter.putUserData(PyConsoleCopyHandler.PROMPT_LENGTH_MARKER, prompt.length() + 1);
+  }
+
+  @NotNull
+  protected String addTextRangeToHistory(@NotNull TextRange textRange, @NotNull EditorEx inputEditor, boolean preserveMarkup) {
```

Comment:
Yes, it is a copy paste with some modification.  I haven't found a way to avoid copying without doing things **worse than copying**.  I consider labyrinthine inheritance hierarchies worse than copying, for example. YYMV.